test/core/libp2p/crypto/key_generator_test.cpp

```diff
@@ -0,0 +1,181 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "libp2p/crypto/key_generator/key_generator_impl.hpp"
+
+#include <gtest/gtest.h>
+#include <openssl/pem.h>
+#include <openssl/rsa.h>
+#include <boost/filesystem.hpp>
+#include <gsl/gsl_util>
+#include "libp2p/crypto/error.hpp"
+#include "libp2p/crypto/random_generator/boost_generator.hpp"
+#include "testutil/outcome.hpp"
+
+using kagome::common::Buffer;
+using libp2p::crypto::Key;
+using libp2p::crypto::KeyGeneratorError;
+using libp2p::crypto::KeyGeneratorImpl;
+using libp2p::crypto::common::RSAKeyType;
+using libp2p::crypto::random::BoostRandomGenerator;
+
+
+class KeyTest : public virtual testing::Test {
+ protected:
+  BoostRandomGenerator random_;
+  KeyGeneratorImpl keygen_{random_};
+};
+
+class GenerateKeyTest : public KeyTest {
+ public:
+  /**
+   * @brief tests generating key pair against specified key type
+   * @param key_type specified key type
+   */
+  void testGenerateKey(Key::Type key_type) {
+    EXPECT_OUTCOME_TRUE_2(val, keygen_.generateKeys(key_type));
+    ASSERT_EQ(val.privateKey.type, key_type);
+    ASSERT_EQ(val.publicKey.type, key_type);
+  }
+
+  /**
+   * @brief ensures that 2 generated key pairs of the same type
+   * are different
+   * @param key_type specified key type
+   */
+  void testKeysAreDifferent(Key::Type key_type) {
+    EXPECT_OUTCOME_TRUE_2(val1, keygen_.generateKeys(key_type));
+    EXPECT_OUTCOME_TRUE_2(val2, keygen_.generateKeys(key_type));
+    ASSERT_NE(val1.privateKey.data.toVector(), val2.privateKey.data.toVector());
+    ASSERT_NE(val1.publicKey.data.toVector(), val2.privateKey.data.toVector());
+  }
+};
+
+class DeriveKeyTest : public KeyTest {
+ protected:
+  void testDerivePublicKey(Key::Type key_type) {
+    EXPECT_OUTCOME_TRUE_2(keys, keygen_.generateKeys(key_type));
+    EXPECT_OUTCOME_TRUE_2(derived, keygen_.derivePublicKey(keys.privateKey));
+    ASSERT_EQ(derived.type, key_type);
+    ASSERT_EQ(keys.publicKey.data.toVector(), derived.data.toVector());
+  }
+};
+
+/**
+ * @given key generator
+ * @when generateRsa of type RSA1024 is called
+ * @then obtained key pair has type RSA1024
+ */
+TEST_F(GenerateKeyTest, GenerateRSA1024Success) {
+  testGenerateKey(Key::Type::RSA1024);
```

Comment:
Use value parameterized tests instead. Also do not put asserts in another function in test. This leads to badly formed messages when such assertions fail

Reply:
done